### DISCUSSION TITLE
Packaging for distribution

### DISCUSSION THREAD

**Discussion by bicarlsen** [extendr/extendr#530] [Q&A]
Is there a guide for how to package our library for distribution, whether locally or on CRAN? Is there an automated process for this (similar to [`maturin`](https://www.maturin.rs/distribution.html)), or do we need to package it ourselves?

If ourselves, which files need to be included?

**Comment by Ilia-Kosenkov** [comment]
Can you explain a little bit what do you mean by 'package'? If you are talking about R package that uses `extendr`, then you have basically two options -- distribute it as is and require all users to have `cargo` and `rust` installed, along with `R` and `Rtools` on windows, or you can try including pre-computed binaries, which may survive CRAN checks, though I am not entirely sure. @yutannihilation has a lot of experience in this area.

**Reply by bicarlsen** [reply]
Yes, we are interested in submitting our package to CRAN, so it would be great to understand what needs to be done to compose everything for submission.

**Reply by CGMossa** [reply]
An alternative is R-universe, which are more willing to accommodate Rust packages. For now, we have to employ diplomacy.

**Reply by eitsupi** [reply]
I think just use the template generated by the edge version of [rextendr](https://github.com/extendr/extendr).
I believe that the template can be used to create packages that can be submitted to CRAN like prqlr.
https://github.com/eitsupi/prqlr

However, be careful with the version of Rust available on the CRAN builder.

**Reply by bicarlsen** [reply]
@eitsupi, what do you mean by the `edge` version of rextendr? And how can we see the version of rust available on CRAN?

**Reply by eitsupi** [reply]
> what do you mean by the edge version of rextendr?

I mean the latest development version on GitHub.

> how can we see the version of rust available on CRAN?

Check the Rust version available for installation with the package manager of the Linux distribution, which can be found here.
https://cran.r-project.org/web/checks/check_flavors.html

The Debian one should be the oldest version. It is currently 1.63.
https://tracker.debian.org/pkg/rustc

I also recommend reading this repository.
https://github.com/r-rust/faq
